app/models/account.rb

```diff
@@ -208,4 +212,25 @@ def self.ransackable_scopes(_auth_object = nil)
       search_for ordered_by
     ]
   end
+
+  private
+
+  def check_associated_records
+    unless rate_management_projects.none?
+      ids = rate_management_projects.pluck(:id)
+      errors.add(:base, "Can't be deleted because linked to Rate Management Project(s) ##{ids.join(', #')}")
+      throw(:abort)
+    end
+
+    unless rate_management_pricelist_items.none?
+      pricelist_ids = rate_management_pricelist_items.pluck(:pricelist_id).uniq
+      pricelists = RateManagement::Pricelist.where(id: pricelist_ids)
+      if pricelists.pluck(:state_id).uniq.all?(RateManagement::PricelistState::CONST::STATE_ID_APPLIED)
+        rate_management_pricelist_items.update_all(account_id: nil)
+      else
+        errors.add(:base, "Can't be deleted because linked to Rate Management Pricelist(s) ##{pricelist_ids.join(', #')}")
+        throw(:abort)
+      end
+    end
+  end
```

Comment:
to many code here. it code be done more elegant and clean:

```ruby
has_many :rate_management_projects, class_name: 'RateManagement::Project'

has_many :active_rate_management_pricelist_items, 
    lambda { not_applied }, 
    class_name: 'RateManagement::PricelistItem'

has_many :applied_rate_management_pricelist_items, 
    lambda { applied }, 
    class_name: 'RateManagement::PricelistItem',
    dependent: :nullify

...

def check_associated_records
  project_ids = rate_management_projects.pluck(:id)
  if project_ids.any?
    errors.add(:base, "Can't be deleted because linked to Rate Management Project(s) ##{project_ids.join(', #')}")
    throw(:abort)
  end

  item_ids = active_rate_management_pricelist_items.pluck(Arel.sql('DISTINCT(pricelist_id)'))
  if item_ids.any?
    errors.add(:base, "Can't be deleted because linked to not applied Rate Management Pricelist(s) ##{project_ids.join(', #')}")
    throw(:abort)
  end
end
```

Reply:
same for below